src/persist/src/operators/stream.rs

```diff
@@ -303,6 +308,57 @@ where
     }
 }
 
+/// Extension trait for [`Stream`].
+pub trait AwaitFrontier<G: Scope<Timestamp = u64>, D> {
+    /// Stashes data until it is no longer beyond the input frontier.
+    ///
+    /// This is similar, in spirit, to what `consolidate()` does for differential collections and
+    /// what `delay()` does for timely streams. However, we are not working with differential
+    /// collections and `delay()` deals with changing the timestamp while the behaviour we want is to
+    /// wait for the frontier to pass. The latter is an implementation detail of `delay()` that is not
+    /// advertised in its documentation.
+    fn await_frontier(&self, name: &str) -> Stream<G, (D, u64, isize)>;
+}
+
+impl<G, D> AwaitFrontier<G, D> for Stream<G, (D, u64, isize)>
+where
+    G: Scope<Timestamp = u64>,
+    D: TimelyData,
+{
+    // Note: This is mostly a copy of the timely delay() operator without the delaying part.
+    fn await_frontier(&self, name: &str) -> Stream<G, (D, u64, isize)> {
+        let operator_name = format!("await_frontier({})", name);
+
+        let mut elements = HashMap::new();
+
+        self.unary_notify(
+            Pipeline,
+            &operator_name,
+            vec![],
+            move |input, output, notificator| {
+                input.for_each(|time, data| {
+                    elements
+                        .entry(time.clone())
+                        .or_insert_with(|| {
+                            notificator.notify_at(time.retain());
+                            Vec::new()
+                        })
+                        .push(data.replace(Vec::new()));
+                });
+
+                // For each available notification, send corresponding set.
+                notificator.for_each(|time, _, _| {
```

Comment:
so this is me not fully understanding the timely notificator api but can we get notifications for times that we didn't `notify_at`? If not, the call to `remove` should always return `Some(...)` right or am I missing something?